src/function/gds/K_Core_Decomposition.cpp

```diff
@@ -0,0 +1,324 @@
+    }
+
+    void pinVertexValues(common::table_id_t tableID) {
+        vertexValues = curVertexValues.getData(tableID);
+    }
+
+    void beginFrontierComputeBetweenTables(table_id_t curTableID, table_id_t nextTableID) override {
+        FrontierPair::beginFrontierComputeBetweenTables(curTableID, nextTableID);
+        pinVertexValues(nextTableID);
+    }
+
+    void beginNewIterationInternalNoLock() override {
+        std::swap(curDenseFrontier, nextDenseFrontier);
+        updateSmallestDegree();
+        curDenseFrontier->ptrCast<PathLengths>()->incrementCurIter();
+        nextDenseFrontier->ptrCast<PathLengths>()->incrementCurIter();
+    }
+
+    uint64_t addToVertexDegree(common::nodeID_t nodeID, uint64_t degreeToAdd) {
+        return curVertexValues.getData(nodeID.tableID)[nodeID.offset].fetch_add(degreeToAdd,
+            std::memory_order_relaxed);
+    }
+
+    uint64_t removeFromVertex(common::nodeID_t nodeID) {
+        int curSmallest = curSmallestDegree.load(std::memory_order_relaxed);
+        int curVertexDegree =
+            curVertexValues.getData(nodeID.tableID)[nodeID.offset].load(std::memory_order_relaxed);
+        if (curVertexDegree > curSmallest) {
+            return curVertexValues.getData(nodeID.tableID)[nodeID.offset].fetch_sub(1,
+                std::memory_order_relaxed);
+        }
+        return curVertexDegree;
+    }
+
+    void updateSmallestDegree(bool dontCheckOffset = false) {
+        if (updateDegreeFlag) {
+            updateDegreeFlag = false;
+            return;
+        }
+        uint64_t curSmallest = UINT64_MAX;
+        for (const auto& [tableID, curNumNodes] : numNodesMap) {
+            curDenseFrontier->pinTableID(tableID);
+            for (uint64_t offset = 0; offset < curNumNodes; ++offset) {
+                if (dontCheckOffset || curDenseFrontier->isActive(offset)) {
+                    curSmallest = std::min(curSmallest,
+                        curVertexValues.getData(tableID)[offset].load(std::memory_order_relaxed));
+                }
+            }
+        }
+        uint64_t current = curSmallestDegree.load(std::memory_order_relaxed);
+        curSmallestDegree.compare_exchange_strong(current, curSmallest, std::memory_order_relaxed);
+    }
+
+    void setAllActive() {
+        for (const auto& [tableID, curNumNodes] : numNodesMap) {
+            curDenseFrontier->pinTableID(tableID);
+            for (uint64_t offset = 0; offset < curNumNodes; ++offset) {
+                common::nodeID_t nodeID;
+                nodeID.tableID = tableID;
+                nodeID.offset = offset;
+                curDenseFrontier->setActive(nodeID);
+            }
+        }
+    }
+
+    uint64_t getSmallestDegree() { return curSmallestDegree.load(std::memory_order_relaxed); }
+
+    void setUpdateFlag() { updateDegreeFlag = true; }
+
+    bool isNodeActive(common::nodeID_t nodeID) const {
+        return curDenseFrontier->isActive(nodeID.offset);
```

Comment:
again why not use isActive interface in GDSFrontier and yet expose an identical one in this class?